packages/block-library/src/table-of-contents/index.php

```diff
@@ -6,28 +6,219 @@
+			$slug  = $block['attrs']['slug'] ?? '';
+			$theme = $block['attrs']['theme'] ?? '';
+
+			if ( $slug ) {
+				$template_part = null;
+
+				if ( function_exists( 'get_block_template' ) ) {
+					$template_id   = $theme ? $theme . '//' . $slug : get_stylesheet() . '//' . $slug;
+					$template_part = get_block_template( $template_id, 'wp_template_part' );
+				}
+
+				if ( $template_part && ! empty( $template_part->content ) ) {
+					$template_blocks   = parse_blocks( $template_part->content );
+					$template_headings = block_core_table_of_contents_traverse_blocks( $template_blocks, $max_level );
+					$headings          = array_merge( $headings, $template_headings );
+				}
+			}
+			continue;
+		}
+
+		// Extract headings from heading blocks.
+		if ( 'core/heading' === $block['blockName'] ) {
+			// Default to level 2 if not specified.
+			$level = isset( $block['attrs']['level'] ) ? $block['attrs']['level'] : 2;
+
+			// Skip if level exceeds max level.
+			if ( $max_level && $level > $max_level ) {
+				continue;
+			}
+
+			$content = '';
+			if ( isset( $block['innerHTML'] ) ) {
+				$content = wp_strip_all_tags( $block['innerHTML'] );
+			} elseif ( isset( $block['innerContent'][0] ) ) {
+				$content = wp_strip_all_tags( $block['innerContent'][0] );
+			}
+
+			if ( empty( trim( $content ) ) ) {
+				continue;
+			}
+
+			// Get anchor from block attributes, or generate one from content.
+			$anchor = isset( $block['attrs']['anchor'] ) ? $block['attrs']['anchor'] : '';
+			if ( empty( $anchor ) && function_exists( 'sanitize_title' ) ) {
+				$anchor = sanitize_title( $content );
+			}
+
+			$headings[] = array(
+				'content' => $content,
+				'level'   => $level,
+				'link'    => ! empty( $anchor ) ? '#' . $anchor : '',
+			);
+		}
+
+		if ( ! empty( $block['innerBlocks'] ) ) {
+			$inner_headings = block_core_table_of_contents_traverse_blocks( $block['innerBlocks'], $max_level );
+			$headings       = array_merge( $headings, $inner_headings );
+		}
+	}
+
+	return $headings;
+}
+
+/**
+ * Build the HTML for the table of contents list.
+ *
+ * @param array $headings Array of heading data.
+ * @param bool  $ordered  Whether to use ordered list.
+ *
+ * @return string HTML for the list.
+ */
+function block_core_table_of_contents_build_list( $headings, $ordered ) {
+	if ( empty( $headings ) ) {
+		return '';
+	}
+
+	$html          = '';
+	$list_tag      = $ordered ? 'ol' : 'ul';
+	$prev_level    = null;
+	$nesting_depth = 0;
+
+	foreach ( $headings as $index => $heading ) {
+		$level = $heading['level'];
+
+		if ( null === $prev_level ) {
+			$prev_level = $level;
+		}
+
+		$depth_change = $level - $prev_level;
+
+		if ( 0 > $depth_change ) {
+			for ( $i = 0; $i < abs( $depth_change ); $i++ ) {
+				$html .= '</li></' . $list_tag . '>';
+				--$nesting_depth;
+			}
+			$html .= '</li>';
+		} elseif ( $depth_change > 0 ) {
+			for ( $i = 0; $i < $depth_change; $i++ ) {
+				$html .= '<' . $list_tag . '>';
+				++$nesting_depth;
```

Comment:
Is this the same as `$nesting_depth++`? I've not seen it at the start before...

Reply:
Oh, apparently it breaks phpcs 😄

`Stand-alone post-increment statement found. Use pre-increment instead: ++$nesting_depth.`

Reply:
TIL!